### PR TITLE
[user-authn] add OIDC support to basic auth proxy

### PR DESCRIPTION
**Pull request by nabokihms**
## Description
Our OIDC Basic Auth Proxy facilitates user authentication within Kubernetes API using basic authentication, bridging the gap between users and the API with enhanced security measures.

## Why do we need it, and what problem does it solve?
Kubernetes API lacks native support for OIDC authentication, posing a challenge for integrating with existing identity providers seamlessly. Our OIDC Basic Auth Proxy resolves this issue by intercepting basic authentication requests, validating user credentials, and authenticating users against the OIDC provider.

## What is the expected result?
It is possible to use OIDC for authentication with basicAuth. 

```
root@dev-master-0:~# cat dex-provider.yaml 
apiVersion: deckhouse.io/v1
kind: DexProvider
metadata:
  name: test
spec:
  type: OIDC
  displayName: test
  oidc:
    issuer: http://dex.default:5556
    clientID: example-app
    clientSecret: ZXhhbXBsZS1hcHAtc2VjcmV0
    insecureSkipEmailVerified: true
    enableBasicAuth: true
```

```
stefan@Stefans-MacBook-Pro ~ % cat kubeconf 
apiVersion: v1
clusters:
- cluster:
    certficate-authority-data: ca.crt
    server: https://api.test-cluster
  name: kubernetes
contexts:
- context:
    cluster: kubernetes
    user: test
  name: test@kubernetes
current-context: test@kubernetes
kind: Config
preferences: {}
users:
- name: test
  user:
    username: admin@example.com
    password: password
```

```
stefan@Stefans-MacBook-Pro ~ % kubectl get po --kubeconfig kubeconf --insecure-skip-tls-verify
Error from server (Forbidden): pods is forbidden: User "admin@example.com" cannot list resource "pods" in API group "" in the namespace "default"
```

```
root@dev-master-0:~# kubectl logs -n d8-user-authn basic-auth-proxy-5dff85d896-n7gfr 
2024-05-17 10:07:54.775718 I | proxy: -- Kubernetes API URL: https://kubernetes.default
2024-05-17 10:07:54.775726 I | proxy: -- Auth Cache TTL: 10s
2024-05-17 10:07:54.775729 I | proxy: -- Groups Cache TTL: 2m0s
--------------------------------
[ Starting Basic auth proxy ]
--------------------------------
2024-05-17 10:07:54.783422 I | proxy: -- OIDC URL: http://dex.default:5556
2024-05-17 10:08:06.132421 I | proxy: receive a new request
2024-05-17 10:08:06.132463 I | oidc provider: validate credentials
2024-05-17 10:08:06.216253 I | oidc provider: validate credentials successful
2024-05-17 10:08:06.216294 I | proxy: received groups for admin@example.com: []
```

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: user-authn
type: feature
summary: Add OIDC support to basic auth proxy.
```

### CODE REVIEW

File: modules/150-user-authn/images/basic-auth-proxy/app/pkg/proxy/provider/crowd.go [hunk 1]
```diff
@@ -127,3 +128,29 @@ func (c *CrowdClient) GetGroups(body string) ([]string, error) {
 	}
 	return groups, nil
 }
+
+type CrowdProvider struct {
+	client *crowdClient
+}
+
+func (c *CrowdProvider) ValidateCredentials(login, password string) ([]string, error) {
+	_, err := c.client.MakeRequest("/session", "POST", struct {
```
Comment (yalosev) [inline modules/150-user-authn/images/basic-auth-proxy/app/pkg/proxy/provider/crowd.go:137]:
```suggestion
	_, err := c.client.MakeRequest("/session", http.MethodPost, struct {
```

File: modules/150-user-authn/images/basic-auth-proxy/app/pkg/proxy/provider/crowd.go [hunk 2]
```diff
@@ -127,3 +128,29 @@ func (c *CrowdClient) GetGroups(body string) ([]string, error) {
...
+		Username string `json:"username"`
+		Password string `json:"password"`
+	}{Username: login, Password: password})
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := c.client.MakeRequest("/user/group/nested?username="+login, "GET", nil)
```
Comment (yalosev) [inline modules/150-user-authn/images/basic-auth-proxy/app/pkg/proxy/provider/crowd.go:145]:
```suggestion
	body, err := c.client.MakeRequest("/user/group/nested?username="+login, http.MethodGet, nil)
```

File: modules/150-user-authn/images/basic-auth-proxy/app/pkg/proxy/provider/oidc.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: (&net.Dialer{
+				Timeout:   30 * time.Second,
+				KeepAlive: 30 * time.Second,
+			}).DialContext,
+			MaxIdleConns:          10,
+			IdleConnTimeout:       30 * time.Second,
+			TLSHandshakeTimeout:   10 * time.Second,
+			ExpectContinueTimeout: 1 * time.Second,
+			TLSClientConfig:       &tls.Config{InsecureSkipVerify: true},
```
Comment (yalosev) [inline modules/150-user-authn/images/basic-auth-proxy/app/pkg/proxy/provider/oidc.go:55]: Isn't `InsecureSkipVerify` potentially risky. Looks possible to have MITM here. Is it for custom CA?

File: modules/150-user-authn/images/basic-auth-proxy/app/pkg/proxy/provider/oidc.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+		},
+	}
+
+	groups := make(map[string]struct{})
```
Comment (yalosev) [inline modules/150-user-authn/images/basic-auth-proxy/app/pkg/proxy/provider/oidc.go:59]:
```suggestion
	groups := make(map[string]struct{}, len(allowedGroups))
```

File: modules/150-user-authn/images/basic-auth-proxy/app/pkg/proxy/provider/oidc.go (hunk 3 above)
Comment (nabokihms) [reply modules/150-user-authn/images/basic-auth-proxy/app/pkg/proxy/provider/oidc.go:55]: This is a copy paste from the crowd provider. Probably, it is best to take the CA option from the provider, but not in the first round of implementation.

File: modules/010-user-authn-crd/crds/doc-ru-dex-provider.yaml [hunk 5]
```diff
@@ -200,6 +200,15 @@ spec:
                     insecureSkipVerify:
                       description: |
                         Не производить проверку подлинности провайдера с помощью TLS. Небезопасно, не рекомендуется использовать в production-окружениях.
+                    enableBasicAuth:
+                      description: |
+                        Включает возможность basic-авторизации для Kubernetes API server.
+
+                        В качестве credentials для basic-авторизации указываются логин и пароль пользователя из приложения, созданного в OIDC (возможно включить при указании только одного провайдера с типом OIDC/Crowd).
+
+                        Работает **только** при включенном `publishAPI`.
```
Comment (z9r5) [inline modules/010-user-authn-crd/crds/doc-ru-dex-provider.yaml:209]:
```suggestion
                        Работает **только** при включенном параметре [publishAPI](https://deckhouse.ru/documentation/v1/modules/150-user-authn/configuration.html#parameters-publishapi).
```

File: modules/010-user-authn-crd/crds/dex-provider.yaml [hunk 6]
```diff
@@ -302,6 +302,15 @@ spec:
                         If a custom certificate isn't provided, this option can be used to turn off
                         TLS certificate checks. As noted, it is insecure and shouldn't be used outside
                         of explorative phases.
+                    enableBasicAuth:
+                      type: boolean
+                      description: |
+                        Enables basic authorization for the Kubernetes API server.
+
+                        The username and password of the user from the application created in OIDC are used as credentials for basic authorization (you can enable it only if there is just one provider of the OIDC/Crowd type).
+                        Works **only** if the `publishAPI` is enabled.
```
Comment (z9r5) [inline modules/010-user-authn-crd/crds/dex-provider.yaml:311]:
```suggestion
                        Works **only** if the [publishAPI](https://deckhouse.io/documentation/v1/modules/150-user-authn/configuration.html#parameters-publishapi) parameter is enabled.
```

**Review by z9r5 — APPROVED** [review]

**Comment by raabdullaev** [comment]
@nabokihms please review.

**Review by yalosev — APPROVED** [review]

**Pull request merged**